### PR TITLE
Prune the heap

### PR DESCRIPTION
**Pull request by litghost**
#### Description

VPR currently has unbounded memory usage as a result of #1032.  This PR caps the heap memory usage at 4x the number of nodes in the entire graph.  This is likely overkill for non-RCV routers.  For RCV based routers, this pruning will be problematic, but there likely needs to be a bounded memory limit for RCV based routers too.

#### Related Issue

#1032

#### Motivation and Context

The current heap implementations (both BinaryHeap and Bucket) can consume an unbounded amount of memory if the router is exploring bad paths that have incrementally better costs.  This has been observed to allow memory usage to grow to consume the entire RAM of ~128 GB machines in pathological cases.  Based on analysis, this memory usage around ~1000 heap entries for each node in the graph.

In #1032 other optimizations were identified, such as updating the best cost to node during insertion, rather than node pop, but even in that case the pathological behavior can result in unbounded memory usage.  This PR bounds the memory usage even in that pathological case.

#### How Has This Been Tested?

- [ ] Exists tests pass
- [ ] Failing unbounded memory case no longer consumes unbounded memory

#### Types of changes

- [x] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed

### CODE REVIEW

File: vpr/src/route/heap_type.h [hunk 1]
```diff
@@ -191,6 +191,27 @@ class HeapInterface {
     // Note: Only invoke this method if all objects returned from this
     // HeapInterface instace have been free'd.
     virtual void free_all_memory() = 0;
+
+    // Set maximum number of elements that the heap should contain
+    // (the prune_limit).  If the prune limit is hit, then the heap should
+    // kick out duplicate index entries.
+    //
+    // For non-RCV routers this simply means taking the lowest cost entry for
+    // each index.  For RCV routers this is ???.
```
Comment (vaughnbetz) [inline vpr/src/route/heap_type.h:200]: For RCV routers we still want the lowest cost item in the heap (the router cost function is more complex, but we still want the lowest cost one).  So you can update or delete this comment (always want to keep the lowest cost items).

Comment (vaughnbetz) [reply vpr/src/route/heap_type.h:200]: I think it would be good to give some context of why this is necessary: basically what your wrote in the issue (can get many slightly better paths in some architectures, leading to the same node being on the heap many times -- if this happens a lot, memory usage can get out of control so we prune and save only the lowest cost entry on the heap for each node).

**Comment by vaughnbetz** [comment]
Code looks good; just requesting a couple more comments.  Should put in the qor comparisons on the vtr and titan benchmarks too (not just regtests passed) make sure we haven't degraded the achieved wiring or the router runtime.  I don't expect much change in either, but good to check now.

File: vpr/src/route/heap_type.h (hunk 1 above)
Comment (litghost) [reply vpr/src/route/heap_type.h]: Awesome, I've updated the comment.  Good to know that this will generalize to the RCV router too!

**Comment by litghost** [comment]
I've pushed the latest version with some comment and code fixes.  I'll post the QoR results once the "Nightly Tests" completes.

**Comment by litghost** [comment]
Found and fixed a bug in the `binary_heap` compaction logic.  I'll pull QoR tommorow and post a comparision.

**Comment by litghost** [comment]
QoR data looks good: [comparison_output.xlsx](https://github.com/verilog-to-routing/vtr-verilog-to-routing/files/4995975/comparison_output.xlsx)

**Comment by litghost** [comment]
Don't merge just yet, I think I found a missing edge case in an overnight test.  Looking now.

**Comment by vaughnbetz** [comment]
Yup, QoR data looks good (probably pruning never kicked in, which is good).
Would also be good to do a stress test on at least one circuit where you set the prune size to be small (num_rr_nodes?  0.25 * num_rr_nodes?) and print a message every time the heap is pruned to ensure the pruning code has been exercised and doesn't crash or cause a route failure.

**Comment by litghost** [comment]
> Yup, QoR data looks good (probably pruning never kicked in, which is good).
> Would also be good to do a stress test on at least one circuit where you set the prune size to be small (num_rr_nodes? 0.25 * num_rr_nodes?) and print a message every time the heap is pruned to ensure the pruning code has been exercised and doesn't crash or cause a route failure.

That's what I've been testing on.  The binary_heap pruning logic worked fine, but I got an assertion failure on the the bucket logic.  FYI, the binary_heap compacted a handful of times (5-6).

**Comment by litghost** [comment]
> Yup, QoR data looks good (probably pruning never kicked in, which is good).
> Would also be good to do a stress test on at least one circuit where you set the prune size to be small (num_rr_nodes? 0.25 * num_rr_nodes?) and print a message every time the heap is pruned to ensure the pruning code has been exercised and doesn't crash or cause a route failure.

Just for reference, pruning 21 million binary heap elements takes ~1 second per prune.  So as long as the prunes are infrequenct, the overall impact is a small CPU hit, but bounded memory usage.

**Comment by litghost** [comment]
@vaughnbetz Go ahead and merge, the assertion error I saw is unrelated to this PR.
